src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/AbstractLanguageServer.cs

```diff
@@ -203,21 +218,93 @@ private void CheckServerState()
         }
     }
 
-    public async Task ShutdownAsync(string message = "Shutting down")
+    public async Task WaitForExitAsync()
+    {
+        lock (_lifeCycleLock)
+        {
+            // Ensure we've actually been asked to shutdown before waiting.
+            if (_shutdownRequestTask == null)
+            {
+                throw new InvalidOperationException("The language server has not yet been asked to shutdown.");
+            }
+        }
+
+        // Note - we wait for the _serverExitedSource task here instead of the _exitNotification task as we may not have
+        // finished processing the exit notification before a client calls into us asking to restart.
+        // This is because unlike shutdown, exit is a notification where clients do not need to wait for a response.
+        await _serverExitedSource.Task.ConfigureAwait(false);
+    }
+
+    /// <summary>
+    /// Tells the LSP server to stop handling any more incoming messages (other than exit).
+    /// Typically called from an LSP shutdown request.
+    /// </summary>
+    public Task ShutdownAsync(string message = "Shutting down")
+    {
+        Task shutdownTask;
+        lock (_lifeCycleLock)
+        {
+            // Run shutdown or return the already running shutdown request.
+            _shutdownRequestTask ??= Shutdown_NoLockAsync(message);
```

Comment:
The reality is that multiple sources could try and tell the server to close for different reasons at the same time.  For example
1.  Request execution errors processing a mutating request
2.  StreamJsonRpc disconnect event
3.   LSP client triggers shutdown/exit (e.g. sln close).

Trying to get all of those sources to play nice in an ordered way or making shutdown/exit logic always handle double requests is imho a recipe for disaster.  Instead we just ensure that shutdown/exit runs once, further calls just get the result.